src/index.ts

```diff
@@ -1,53 +1,266 @@
+    requirejs.require.config({
+      baseUrl: baseURL
+    });
   }
 
-  // Finally, we allow returning a promise (or an async function above).
-  plugin = await Promise.resolve(plugin);
+  private async _loadPlugin(code: string, path: string | null) {
+    const tokenMap = new Map(
+      Array.from((this.app as any)._serviceMap.keys()).map((t: any) => [
+        t.name,
+        t
+      ])
+    );
+    // Widget registry does not follow convention of importName:tokenName
+    tokenMap.set(
+      '@jupyter-widgets/base:IJupyterWidgetRegistry',
+      tokenMap.get('jupyter.extensions.jupyterWidgetRegistry')
+    );
+    const importResolver = new ImportResolver({
+      modules: modules as unknown as Record<string, IModule>,
+      tokenMap: tokenMap,
+      requirejs: this.requirejs,
+      settings: this.settings,
+      serviceManager: this.app.serviceManager,
+      basePath: path
+    });
+
+    const pluginLoader = new PluginLoader({
+      transpiler: new PluginTranspiler({
+        compilerOptions: {
+          target: ts.ScriptTarget.ES2017,
+          jsx: ts.JsxEmit.React
+        }
+      }),
+      importFunction: importResolver.resolve.bind(importResolver),
+      tokenMap: tokenMap,
+      serviceManager: this.app.serviceManager,
+      requirejs: this.requirejs
+    });
+    importResolver.dynamicLoader = pluginLoader.loadFile.bind(pluginLoader);
+
+    let result;
+    try {
+      result = await pluginLoader.load(code, path);
+    } catch (error) {
+      if (error instanceof PluginLoadingError) {
+        const internalError = error.error;
+        showDialog({
+          title: `Plugin loading failed: ${internalError.message}`,
+          body: formatErrorWithResult(error, error.partialResult)
+        });
+      } else {
+        showErrorMessage('Plugin loading failed', (error as Error).message);
+      }
+      return;
+    }
+    const plugin = result.plugin;
 
-  plugin.requires = plugin.requires?.map((value: any) => token_map.get(value));
-  plugin.optional = plugin.optional?.map((value: any) => token_map.get(value));
+    if (result.schema) {
+      // TODO: this is mostly fine to get the menus and toolbars, but:
+      // - transforms are not applied
+      // - any refresh from the server might overwrite the data
+      // - it is not a good long term solution in general
+      this.settingRegistry.plugins[plugin.id] = {
+        id: plugin.id,
+        schema: JSON.parse(result.schema),
+        raw: result.schema,
+        data: {
+          composite: {},
+          user: {}
+        },
+        version: '0.0.0'
+      };
+      (
+        this.settingRegistry.pluginChanged as Signal<ISettingRegistry, string>
+      ).emit(plugin.id);
+    }
 
-  (app as any).registerPluginModule(plugin);
-  if (plugin.autoStart) {
-    await app.activatePlugin(plugin.id);
+    // Unregister plugin if already registered.
+    if (this.app.hasPlugin(plugin.id)) {
+      delete (this.app as any)._pluginMap[plugin.id];
```

Comment:
The leaf/dependent plugins are not unloaded (see https://github.com/jupyterlab/lumino/issues/278#issuecomment-1007323877). I guess this is ok for now